app/scripts/lib/account-tracker.ts

```diff
@@ -457,35 +546,37 @@ export default class AccountTracker {
    * balanceChecker is deployed on main eth (test)nets and requires a single call
    * for all other networks, calls this.#updateAccount for each account in this.store
    *
-   * @param {string} networkClientId - optional network client ID to use instead of the globally selected network.
-   * @returns {Promise} after all account balances updated
+   * @param networkClientId - optional network client ID to use instead of the globally selected network.
    */
-  async updateAccounts(networkClientId) {
-    const { completedOnboarding } = this.onboardingController.state;
+  async updateAccounts(networkClientId?: NetworkClientId): Promise<void> {
+    const { completedOnboarding } = this.#onboardingController.state;
     if (!completedOnboarding) {
       return;
     }
 
     const { chainId, provider, identifier } =
       this.#getCorrectNetworkClient(networkClientId);
     const { useMultiAccountBalanceChecker } =
-      this.preferencesController.store.getState();
+      this.#preferencesController.store.getState();
 
     let addresses = [];
     if (useMultiAccountBalanceChecker) {
       const { accounts } = this.store.getState();
 
       addresses = Object.keys(accounts);
     } else {
-      const selectedAddress = this.controllerMessenger.call(
+      const selectedAddress = this.#controllerMessenger.call(
         'AccountsController:getSelectedAccount',
       ).address;
 
       addresses = [selectedAddress];
     }
 
     const rpcUrl = 'http://127.0.0.1:8545';
-    const singleCallBalancesAddress = SINGLE_CALL_BALANCES_ADDRESSES[chainId];
+    const singleCallBalancesAddress =
+      SINGLE_CALL_BALANCES_ADDRESSES[
+        chainId as keyof typeof SINGLE_CALL_BALANCES_ADDRESSES
+      ];
```

Comment:
Let's make sure the type and runtime behavior align here by using a type predicate function instead of a type assertion.

```suggestion
```

Here's the full diff:

```diff
diff --git a/app/scripts/lib/account-tracker.ts b/app/scripts/lib/account-tracker.ts
index 8ca119ccf8..8cb933d799 100644
--- a/app/scripts/lib/account-tracker.ts
+++ b/app/scripts/lib/account-tracker.ts
@@ -573,14 +573,11 @@ export default class AccountTracker {
     }
 
     const rpcUrl = 'http://127.0.0.1:8545';
-    const singleCallBalancesAddress =
-      SINGLE_CALL_BALANCES_ADDRESSES[
-        chainId as keyof typeof SINGLE_CALL_BALANCES_ADDRESSES
-      ];
     if (
       identifier === LOCALHOST_RPC_URL ||
       identifier === rpcUrl ||
-      !singleCallBalancesAddress
+      !((id): id is keyof typeof SINGLE_CALL_BALANCES_ADDRESSES =>
+        id in SINGLE_CALL_BALANCES_ADDRESSES)(chainId)
     ) {
       await Promise.all(
         addresses.map((address) =>
@@ -590,7 +587,7 @@ export default class AccountTracker {
     } else {
       await this.#updateAccountsViaBalanceChecker(
         addresses,
-        singleCallBalancesAddress,
+        SINGLE_CALL_BALANCES_ADDRESSES[chainId],
         provider,
         chainId,
       );
```

Reply:
I will make the change here https://github.com/MetaMask/metamask-extension/pull/27258 as for runtime it used to work this way so there will be no issues on that side

Reply:
@MajorLift I just applied your change here https://github.com/MetaMask/metamask-extension/pull/27258/commits/3b5a9903bc9be85d6dcc9711330aa02ec8768d5c